### ISSUE TITLE
Distinguishing between Multi-class and Multi-label for a Text Classification Problem

### ISSUE THREAD

**Issue by Shea-Fyffe** [flairNLP/flair#2365]
I am attempting to fine-tune a few-shot text classification model with several training examples. _Each training example only has one true label_. 

**I am trying to predict multi-label probabilities (i.e., don't sum up to 1), for unlabeled data. With the example code below my probabilities sum to 1. Is there something that sticks out or a way I can avoid the softmax at the end?**

```

#### Flair for few-shot learning ####
### Import modules
from flair.data import Corpus, Sentence
from flair.datasets import SentenceDataset, ClassificationCorpus
from flair.trainers import ModelTrainer
from flair.models.text_classification_model import TARSClassifier
   
### Helper function
def convert_to_sentence_data(sentences, labels, task_name):
    res = SentenceDataset([
      Sentence(s).add_label(task_name, labels[i]) for i, s in enumerate(sentences)
          ])
    return res

###### Body
## 0. convert example data

# training sentences
train_sent = ['I give my team authority over issues within the department.','I am concerned that my team reach their goal.','I recognize my team strong and weak sides.','I focus on developing human capital (the individual capabilities, knowledge, skills and experience of a firm's employees).','I develop or help develop standard operating procedures and standardized processes.','I clarify task performance strategies.','I act as a representative of the team with other parts of the organization (e.g., other teams, management).','I challenge my team to think about old problems in new ways.','I ensure that the team has a clear understanding of its purpose.','I help provide a clear vision of where the team is going.','I communicate expectations for high team performance.','I define and structures my own work and the work of the team.','I go beyond my own interests for the good of the team.','I ask questions that prompt my employees to think.','I have stimulated my employees to rethink the way they do things.','I treat my employees by considering their personal feelings.']

# each training example has just one label with 5 unique labels in total
train_labels = ['autonomy and empowerment','autonomy and empowerment','autonomy and empowerment','creativity','task management','task management','external representation and boundary management','creativity','task management','task management','task management','task management','personal care and support','creativity','creativity','personal care and support']

# use helper to convert to SentenceDataSet
train_data = convert_to_sentence_data(train_sent, train_labels, "behaviors")


## 1. make corpus
crpus = Corpus(train=train_data)
    
### Run predictions
## 2. load TARS
tars = TARSClassifier(task_name = "behaviors",
                          label_dictionary = crpus.make_label_dictionary(),
                          document_embeddings = "paraphrase-mpnet-base-v2",
                          batch_size = 16, multi_label = True)
   
## 3. initialize the text classifier trainer
    trainer = ModelTrainer(tars, crpus)
    
## 4. train model
    trainer.train(base_path="./example/dir/",
                  learning_rate=0.02,
                  mini_batch_size=8,
                  max_epochs=20,
                  embeddings_storage_mode="gpu")

## 5. load model
    ft_tars = TARSClassifier.load('./example/dir/best-model.pt')
    
## 6. Predict
  test_data = 
   ['i had a conversation  with each receptionist regarding being more efficient in their recording of products in and out',
   'get requests to increase or decrease budgets on projects and to close out projects as they are finished',
   'marketing sales',  'delegated tasks for closing', 
   'i smiled at her and let her know what i would be finished with the reports within 15 minutes',
   'i wanted to make the client comfortable  and i alerted everyone to be professional and be on time',
   'we made sure we had enough expertise and knowledge to show the client that we were the best for business']

    test_sentences = [Sentence(s) for s in test_data]
    ft_tars.predict(test_sentences, multi_class_prob = True)
```

Let me know if there is anything missing!

Thanks for being awesome

**Comment by kishaloyhalder** [comment]
Hi @Shea-Fyffe 

Thanks for raising the issue and sharing your code to reproduce it!

In the current implementation, `TARSClassifier` relies on the `label_dictionary` argument passed in the constructor call to determine whether the classification task is a multi-class or multi-labelling problem. In your case, every text example has a single label and that signals the `TARSClassifier` to treat it as a multi-class problem (with softmax enabled in the end). Consider changing the `make corpus` and `load tars` section in your example to the following to treat it as a multi-labelling problem:

```
#1. make corpus
crpus = Corpus(train=train_data)
label_dict = crpus.make_label_dictionary()
label_dict.multi_label = True

#Run predictions
#2. load TARS
tars = TARSClassifier(task_name = "behaviors",
					label_dictionary = label_dict,
					document_embeddings = "paraphrase-mpnet-base-v2",
					batch_size = 16)
```

Hope this helps. Good luck!

with regards,
Kishaloy

**Comment by alanakbik** [comment]
Closing, but feel free to reopen if you have more questions!

**Comment by Shea-Fyffe** [comment]
A quick clarification on the interpretation of multi-class versus multi-label. I've read some of the original research [(Halder et al., 2020)](https://aclanthology.org/2020.coling-main.285), but I'm not confident in my interpretation. I'm familiar with the `NLI` approach to zero-shot; however, I've read that Flair uses a slightly different architecture [(Peirsman, 2021)](https://nlp.town/blog/zero-shot-classification/)--moreover, I want to make sure this translates to a few-shot task.

Let's say I train the model in a few-shot scenario. In this example I'm only using **2** different classes. *Skip to the end to see my questions*

### 1. Setup analysis & corpus
```
# Flair for few-shot learning
from flair.data import Corpus, Sentence
from flair.datasets import SentenceDataset, ClassificationCorpus
from flair.trainers import ModelTrainer
from flair.models.text_classification_model import TARSClassifier
"""
####
"""
### Helper function
def convert_to_sentence_data(sentences, labels, task_name):
    res = SentenceDataset([
      Sentence(s).add_label(task_name, labels[i]) for i, s in enumerate(sentences)
          ])
    return res

# 17 training sentences --10 initiation of structure --7 consideration
train_sent = ['I let group members know what is expected of them.', 'I am friendly and approachable.', 'I encourage the use of uniform procedures.', 'I do little things to make it pleasant to be a member of the group.', 'I try out my ideas in the group.', 'I put suggestions made by the group into operation.', 'I make my attitudes clear to the group.', 'I treat all group members as my equals.', 'I decide what shall be done and how it shall be done.', 'I give advance notice of changes.', 'I assign group members to particular tasks.', 'I make sure that my part in the group is understood by the group members.', 'I look out for the personal welfare of group members.', 'I schedule the work to be done.', 'I am willing to make changes.', 'I maintain definite standards of performance.', 'I ask that group members to follow standard rules and regulations.']

# each training example is assigned to just 1 of the two labels 
train_labels = ['initiation of structure.', 'consideration.', 'initiation of structure.', 'consideration.', 'initiation of structure.', 'consideration.', 'initiation of structure.', 'consideration.', 'initiation of structure.', 'consideration.', 'initiation of structure.', 'initiation of structure.', 'consideration.', 'initiation of structure.', 'consideration.', 'initiation of structure.', 'initiation of structure.']

# use helper to convert to SentenceDataSet
train_data = convert_to_sentence_data(train_sent, train_labels, "multi_class_behaviors")

# make corpus (without setting multi_label = True)
crpus = Corpus(train=train_data)
label_dict = crpus.make_label_dictionary()
```
**Note:** all of the training examples are positively-keyed so they represent higher levels of the label construct.

### 2. Train a TARS model
For fine-tuning, I'm *not* specifying multi-label
```
#initialize the text classifier and trainer with corpus 
tars = TARSClassifier(task_name = "multi_class_behaviors",
                          label_dictionary = label_dict)

trainer = ModelTrainer(tars, crpus)
    
# train model
trainer.train(base_path="./example/dir/",
                    learning_rate=0.02,
                    mini_batch_size=4,
                    max_epochs=20,
                    embeddings_storage_mode="gpu")

#load model
ft_tars = TARSClassifier.load('./example/dir/best-model.pt')
```

### 3. Format unlabeled test data
```
#load test data
test_data = ['i care more about the work than my employees', 'get requests to increase or decrease budgets on projects and to close out projects as they are finished', 'marketing sales',  'delegated tasks for closing']

test_sentences = [Sentence(s) for s in test_data]
```

### 4a. Predict multi-class labels
```
#predict multi-class probabilities
ft_tars.predict(test_sentences, multi_class_prob = True)
```

### 4b. Predict multi-label probabilities
```
# switch to multi-label task
ft_tars.add_and_switch_to_new_task(
                          task_name = "multi_label_behaviors",
                          label_dictionary = label_dict,
                          multi_label = True,
                          multi_label_threshold = 0.0)
#predict multi-label probabilities
ft_tars.predict(test_sentences)
```

***Questions:*** 

- When interpreting the prediction from section `4a` (i.e., multi-class) I should assume that these values represent the probability that the model thought a test sentence was *related* to `initiating structure` OR `consideration`?
- This is a tricky one (for me). When interpreting the prediction from section `4b` (i.e., multi-label) I should assume that values for `consideration`--for instance--represent the probability that a sentence is *related* to positive consideration? In other words, is this synonymous with entailment in an NLI task? Since my training set only included examples that represent positive consideration and not inconsideration/negative consideration, I must interpret it this way?
- Overall, do the predicted values represent just the strength of a relationship or do they also include direction? So a test case that depicts being *inconsiderate* (e.g., 'i care more about the work than my employees') is likely to receive a lower multi-label predicted value along with test cases that are unrelated to consideration (e.g.,  'marketing sales')?

In summary, my goal is to produce roughly 3 variables for each of the 2 labels (similar to the common `NLI` framework or a even a 3-label sentiment analysis); where one pair are probabilities of a positive/entailment relationship one pair are probabilities of a negative/contradiction relationship, and one pair neutral/unrelated probabilities.

**Comment by Shea-Fyffe** [comment]
> Closing, but feel free to reopen if you have more questions!

I was unsure of how to re-open this issue so I mentioned it in a new yet somewhat related issue #2410
